### PR TITLE
Optimize comparision for collation `UTF8_BIN` and `UTF8MB4_BIN`

### PR DESCRIPTION
**Pull request by solotzg**
### What problem does this PR solve?

Issue Number: ref #5294

Problem Summary:

### What is changed and how it works?

- Reduce virtual function call when handling collaction `UTF8_BIN` and `UTF8MB4_BIN`
- Check tailf space first to reduce branches.
- Optimize comparison related to check equality.

#### Benchmark

Env:
- TPCH 10
- 1 Node
- limit cpu usage up to 200% by cgroup

----

SQL
```
select count(1) from lineitem where l_comment = 'zzle? slyly regular instruc       ';
```

```
MySQL [tpch_10]> select count(1) from lineitem where l_comment = 'zzle? slyly regular instruc       ';
+----------+
| count(1) |
+----------+
|        1 |
+----------+
```

Time(s) | Original | Optimized | NoCollation |   |  
-- | -- | -- | -- | -- | --
  | 2.46 | 1.56 | 1.48 |   |  
  | 2.41 | 1.56 | 1.38 |   |  
  | 2.36 | 1.47 | 1.43 |   |  
  | 2.35 | 1.55 | 1.4 | NoCollation : Original | 69.30%
  | 2.44 | 1.49 | 1.41 | Optimized : Original | 57.54%
AVG | 2.404 | 1.526 | 1.42 | NoCollation : Optimized | 7.46%

----

SQL
```
select count(1) from lineitem where l_comment != l_comment;
```

Time(s) | Original | Optimized | NoCollation |   |  
-- | -- | -- | -- | -- | --
  | 2.37 | 1.66 | 1.96 |   |  
  | 2.25 | 1.52 | 1.9 |   |  
  | 2.32 | 1.66 | 1.94 |   |  
  | 2.24 | 1.63 | 1.94 | NoCollation : Original | 19.27%
  | 2.33 | 1.54 | 1.91 | Optimized : Original | 43.70%
AVG | 2.302 | 1.602 | 1.93 | NoCollation : Optimized | -16.99%

----

SQL
```
select count(1) from lineitem where l_comment < l_comment;
```

Time(s) | Original | Optimized | NoCollation |   |  
-- | -- | -- | -- | -- | --
  | 2.32 | 2.07 | 1.91 |   |  
  | 2.27 | 2.06 | 1.95 |   |  
  | 2.33 | 2.07 | 2.05 |   |  
  | 2.33 | 2.09 | 1.99 | NoCollation : Original | 15.49%
  | 2.23 | 2.08 | 2.04 | Optimized : Original | 10.70%
AVG | 2.296 | 2.074 | 1.988 | NoCollation : Optimized | 4.33%

#### Analyze

When only checking equality, https://github.com/pingcap/tiflash/blob/eaf1a4cf886d1d5a86d04ef14eff956778ab4911/libs/libcommon/include/common/mem_utils.h#L138 performs better than `std::memcmp` if most parts of source and target string are equal.

- maybe because libc is not very new. 

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by windtalker** [comment]
Can you also add some benchmark result for TiDB cluster with new collation disabled?

File: dbms/src/Functions/CollationOperatorOptimized.h [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+{
+    bool has_tail_space = false;
+
+    // #pragma clang loop vectorize(enable)
+    for (size_t i = 0; i < size; ++i)
+    {
+        const auto * ptr = reinterpret_cast<const char *>(&a_data[StringUtil::offsetAt(a_offsets, i)]);
+        auto size = StringUtil::sizeAt(a_offsets, i) - 1;
+        auto pos = size > 0 ? size - 1 : 0; // if size is 0, use the last pos which aways contains a '\0'
+        has_tail_space |= ptr[pos] == ' ';
```
Comment (windtalker) [inline dbms/src/Functions/CollationOperatorOptimized.h:95]: break if `has_tail_space == true`?

File: dbms/src/Functions/CollationOperatorOptimized.h [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+// Compare str view by memcmp
+__attribute__((flatten, always_inline, pure)) inline int RawStrCompare(const std::string_view & v1, const std::string_view & v2)
+{
+    return signum(v1.compare(v2));
+}
+
+// Remove tail space
+__attribute__((flatten, always_inline, pure)) inline std::string_view RightTrim(const std::string_view & v)
+{
+    size_t end = v.find_last_not_of(' ');
```
Comment (fuzhe1989) [inline dbms/src/Functions/CollationOperatorOptimized.h:53]: I wonder if other space characters need to be handled here.

File: dbms/src/Functions/CollationOperatorOptimized.h (hunk 1 above)
Comment (solotzg) [reply dbms/src/Functions/CollationOperatorOptimized.h]: Fixed, refine code in d5dc1b0eced92f188de09a62f2334c51db225b6f and get better performance due to better locality.

File: dbms/src/Functions/CollationOperatorOptimized.h (hunk 2 above)
Comment (solotzg) [reply dbms/src/Functions/CollationOperatorOptimized.h]: No need to consider other cha. According to [MySQL::Doc::string-comparison-functions](https://dev.mysql.com/doc/refman/8.0/en/string-comparison-functions.html).

```
In particular, trailing spaces are always significant. This differs from comparisons performed with the [=](https://dev.mysql.com/doc/refman/8.0/en/comparison-operators.html#operator_equal) operator, for which the significance of trailing spaces in nonbinary strings (CHAR, VARCHAR, and TEXT values) depends on the pad attribute of the collation used for the comparison. For more information, see [Trailing Space Handling in Comparisons](https://dev.mysql.com/doc/refman/8.0/en/charset-binary-collations.html#charset-binary-collations-trailing-space-comparisons).
```

**Comment by solotzg** [comment]
> Can you also add some benchmark result for TiDB cluster with new collation disabled?

Done.

**Review by windtalker — APPROVED** [review]
LGTM

File: dbms/src/Functions/FunctionsComparison.h [hunk 3]
```diff
@@ -720,10 +734,41 @@ class FunctionComparison : public IFunction
...
+                throw Exception("Logical error: ColumnConst contains not String nor FixedString column", ErrorCodes::ILLEGAL_COLUMN);
+        }
+
+        if (c1_const)
+        {
+            if (const auto * c1_const_string = checkAndGetColumn<ColumnString>(&c1_const->getDataColumn()); c1_const_string)
+            {
+                c1_const_str_ref = std::string_view(c1_const_string->getDataAt(0));
+            }
+            else if (const auto * c1_const_fixed_string = checkAndGetColumn<ColumnFixedString>(&c0_const->getDataColumn()); c1_const_fixed_string)
```
Comment (yibin87) [inline dbms/src/Functions/FunctionsComparison.h:760]: should be c1_const->getDataColumn() here? Better to extract a method to do so.

Comment (solotzg) [reply dbms/src/Functions/FunctionsComparison.h]: Fixed. That's a mistake. Thank you very much.

**Review by yibin87 — APPROVED** [review]
LGTM

File: dbms/src/Functions/CollationOperatorOptimized.h [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        size_t a_size = StringUtil::sizeAt(a_offsets, i) - 1;
+        const auto * a_ptr = reinterpret_cast<const char *>(&a_data[StringUtil::offsetAt(a_offsets, i)]);
+
+        func({a_ptr, a_size}, i);
+    }
+}
+
+// Handle str-column compare str-column.
+// - Optimize UTF8_BIN and UTF8MB4_BIN
+//   - Check if columns don NOT contains tail space
```
Comment (zanmato1984) [inline dbms/src/Functions/CollationOperatorOptimized.h:125]:
```suggestion
//   - Check if columns do NOT contain tail space
```

File: dbms/src/Functions/CollationOperatorOptimized.h [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+    default:
+        break;
+    }
+    return use_optimized_path;
+}
+
+// Handle str-column compare const-str.
+// - Optimize UTF8_BIN and UTF8MB4_BIN
+//   - Right trim const-str first
+//   - Check if column don NOT contains tail space
```
Comment (zanmato1984) [inline dbms/src/Functions/CollationOperatorOptimized.h:169]:
```suggestion
//   - Check if column do NOT contain tail space
```

File: dbms/src/Functions/FunctionsComparison.h [hunk 6]
```diff
@@ -342,8 +356,8 @@ struct StringComparisonWithCollatorImpl
     }
 
     static void constantConstant(
-        const std::string & a,
-        const std::string & b,
+        std::string_view a,
```
Comment (zanmato1984) [inline dbms/src/Functions/FunctionsComparison.h:359]: Why not const ref?

Comment (solotzg) [reply dbms/src/Functions/FunctionsComparison.h:359]: `std::string_view` only contain a pointer and a size_t (16 bytes totally). It's fine to use const ref or object itself.

Comment (solotzg) [reply dbms/src/Functions/FunctionsComparison.h]: Fixed

**Review by zanmato1984 — APPROVED** [review]
LGTM

**Pull request merged**